[red-knot] don't include Unknown in the type for a conditionally-defined import

pilleye opened the pull request:
## Summary

Fixes the bug described in #13514 where an unbound public type defaulted to the type or `Unknown`, whereas it should only be the type if unbound.

## Test Plan

Added a new test case

pilleye commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -58,9 +58,7 @@ fn symbol_ty_by_id<'db>(db: &'db dyn Db, scope: ScopeId<'db>, symbol: ScopedSymb
-                use_def
-                    .public_may_be_unbound(symbol)
-                    .then_some(Type::Unknown),
+                use_def.public_may_be_unbound(symbol).then_some(Type::Never),
```
below this is the case where `use_def.has_public_declarations(symbol)` is false, what would be a situation where we fall into that case?

Slyces replied:
```python
if flag:
    x = 1
```
See my comment on #13514

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5452,7 +5486,7 @@ mod tests {
         // TODO: once we support `sys.version_info` branches,
         // we can set `--target-version=py311` in this test
         // and the inferred type will just be `BaseExceptionGroup` --Alex
```
We can now remove this comment, and the similar one in two below tests, because the lack of `sys.version_info` handling no longer results in a spurious `Unknown` here.

carljm reviewed (approved):
Awesome, thank you!! Just a few comment removals and this looks good to go.

carljm commented:
Oh, I didn't notice tests are failing. We'll need to update the tests, and I'll need to look at the resulting assertion changes to make sure we aren't getting any undesired side effects from this.

pilleye commented:
Ah, I missed those tests on the latest version, will fix tonight.

pilleye commented:
There seems to be a weird side-effect where we're narrowing `Unbound | Unknown` types to just `Unknown` instead of `Unbound`: https://github.com/pilleye/ruff/blob/pilleye/public-known/crates/red_knot_python_semantic/src/types/infer.rs#L5313

carljm commented:
So one thing I'm realizing here is that we are using `assert_public_ty` in some tests where we really intend to be testing "what is the local type of this variable at the end of the scope" -- this was natural because those used to be the same thing, but here we are definitively making them different.

So for tests like `basic_for_loop`, for example, what we should really do instead of using `assert_public_type` is to add a `reveal_type(x)` as the last line in the test code, and then `assert_file_diagnostics` that we get the expected type revealed there, which will include `Unbound`.

pilleye commented:
Working through 2 different globals-related bugs right now:

1) We seem to be incorrectly shadowing global builtins with `Never` here: https://github.com/pilleye/ruff/blob/pilleye/public-known/crates/red_knot_python_semantic/src/types/infer.rs#L4431

2) We're not respecting implicit globals: https://github.com/pilleye/ruff/blob/pilleye/public-known/crates/red_knot_python_semantic/src/types/infer.rs#L4425

carljm commented:
Ah ok, so that's going to be due to the use of `global_symbol_ty` in `TypeInferenceBuilter::lookup_name`. We need to use a function there that _does_ union with unbound, so that we can replace unbound with the builtin definition. So the set of functions in `types.rs`: `global_symbol_ty`, `symbol_ty`, `symbol_ty_by_name`, `symbol_ty_by_id` are going to have to be split between the import case and the global-lookup-within-the-module case.

I would probably keep those functions as is, and add a new `global_symbol_lookup` function that a) still inserts `Type::Unbound` in case unbound is possible, and b) should also use only definitions, not declared types, even if there are declarations (so it's like just the second case in `symbol_ty_by_id`). (We may want a new test to verify (b) as well.) We may not even need `global_symbol_lookup` to be a function; I think the only callsite will be in `TypeInferenceBuilder::lookup_name`, so it can probably just be inlined there, since it shouldn't be more than a couple lines.

Let me know if any of that doesn't make sense! Thanks for working through this. I knew the first version of this PR was too simple to be true 😆

pilleye commented:
Haha the first version seemed suspiciously easy

I think I got the tests working, but definitely feels way jankier than I would want. Would appreciate comments!

I think trying to genericize `symbol_ty_by_id` adds unnecessary complexity, so I might just inline that code as you said.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3277,8 +3278,8 @@ mod tests {
 
         // the type as seen from external modules (`Unknown`)
         // is different from the type inside the module itself (`Unbound`):
-        assert_public_ty(&db, "src/package/foo.py", "x", "Unbound");
-        assert_public_ty(&db, "src/package/bar.py", "x", "Unknown");
+        assert_public_ty(&db, "src/package/foo.py", "x", "Never");
+        assert_public_ty(&db, "src/package/bar.py", "x", "Never");
```
I don't think this is a behavior change we want. If you import a fully undeclared/undefined name, you should get a diagnostic, but the imported name should have type `Unknown` in this case. Otherwise you'll just get lots of useless cascading errors about doing invalid things with `Never`.

There is currently a special case in `TypeInferenceBuilder::infer_import_from_definition` where we transform `Unbound` to `Unknown` -- that's why this test exists. That special case becomes obsolete with this PR (we'll never get `Unbound` there anymore), but I think the special case should be updated to check for `Never`, emit a diagnostic (use `self.add_diagnostic`), and transform the `Never` to `Unknown`. (It's also fine if you want to just add a TODO for the diagnostic and we can add it in a later PR; we'll also want a similar diagnostic that's currently missing in `infer_import_definition`, for a nonexistent module.)

pilleye replied:
Just to double-check, `x` as viewed from `bar.py` should be `Unknown` still, not `Never`, right?

carljm replied:
Right, I think we should handle that in the importing code, instead of the current code which replaces unbound with unknown we'll want to replace never with unknown (and emit a diagnostic, but that can be a TODO if you want, doesn't have to be in this PR.)

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4158,6 +4159,40 @@ mod tests {
+    }
+
+    #[test]
+    fn resolve_unannotated_unbound_public() -> anyhow::Result<()> {
```
nit: it's only maybe unbound
```suggestion
    fn resolve_unannotated_maybe_unbound_public() -> anyhow::Result<()> {
```

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4158,6 +4159,40 @@ mod tests {
     }
 
+    #[test]
+    fn resolve_unbound_public() -> anyhow::Result<()> {
```
```suggestion
    fn resolve_maybe_undeclared_maybe_unbound_public() -> anyhow::Result<()> {
```

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4374,8 +4409,8 @@ mod tests {
         assert_public_ty(&db, "src/a.py", "x", "Literal[3, 4, 5]");
-        assert_public_ty(&db, "src/a.py", "r", "Unbound | Literal[2]");
-        assert_public_ty(&db, "src/a.py", "s", "Unbound | Literal[5]");
+        assert_public_ty(&db, "src/a.py", "r", "Literal[2]");
+        assert_public_ty(&db, "src/a.py", "s", "Literal[5]");
```
This is one of those cases where part of the whole purpose of the test is actually to show that we know which names might be unbound, so we should really update the test such that it doesn't use (only)`assert_public_ty`, because it's really not the public type we want to assert on.

I earlier suggested an approach using `reveal_type`, and I think that's actually a fine option in the tests below where you use it.

I realized a simpler approach in some of these tests (particularly the global-scope ones) might be to keep these `assert_public_ty` but add another helper `assert_may_be_unbound` (which uses the use-def map directly) and explicitly assert here that `r` and `s` may be unbound. This way we actually get both the assertion on the public type, and an assertion that locally we do know the name might be unbound.

pilleye replied:
Is there a way to add public type tests to the markdown tests introduced in #13719?

carljm replied:
You can import it (since tests can be multi-file), or refer to the name as a global from within a nested function. Import probably makes it clearer what's going on?

pilleye replied:
How do I run the mdtests? `cargo test` inside `red_knot_python_semantic` and `red_knot_test` both don't seem to run it.

carljm replied:
`cargo test` inside `red_knot_python_semantic` (or `cargo test -p red_knot_python_semantic` from the workspace) should run the mdtests, and do for me. How are you concluding that they aren't running?

pilleye replied:
Hmm, they only run when I run `cargo test mdtest` inside `red_knot_python_semantic`.

Without `mdtest`: https://gist.github.com/pilleye/7454e668bc19f9ca1969e0765a82a301
With `mdtest`: https://gist.github.com/pilleye/3e9842e37a911eb88f6ac26c281bb0dc

carljm replied:
That's weird -- I'm not enough of a `cargo test` expert to explain this discrepancy :) For me they definitely run with just `cargo test`... In general my experience has been that `cargo test` runs all tests it discovers, both unit tests and integration tests. Not sure what might cause that to not be happening for you. `mdtest` is just a normal Rust integration test.

AlexWaygood replied:
huh, that _is_ counterintuitive!

carljm replied:
Ah, that makes sense! TIL

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4516,7 +4551,7 @@ mod tests {
         .unwrap();
 
-        assert_public_ty(&db, "src/a.py", "y", "Unbound | Literal[2, 3]");
+        assert_public_ty(&db, "src/a.py", "y", "Literal[2, 3]");
```
same as above, should have an assertion in some form that `y` can be unbound here

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4632,24 +4667,29 @@ mod tests {
-    #[test]
-    fn conditionally_global_or_builtin() -> anyhow::Result<()> {
+    #[test_case("")]
+    // Tests that we only use the definition of a symbol instead of its declaration when we are
+    // checking module globals without a nonlocal binding.
```
So on further thought and discussion with the team, I think I pointed you in the wrong direction here yesterday; sorry about that! I think we should always use the same logic, respecting both declarations and definitions. (Maybe we will want to consider using only definitions for a not-global nonlocal reference, but that's not a change we should make in this PR.)

I still think the test case you added here is useful! But I would want the assertion to change such that when you add the annotation, you get `Literal[copyright] | int` instead of `Literal[copyright] | Literal[1]`.

And if the assertion changes, it's probably clearer to be a fully separate test rather than a `#[test_case]`?

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5360,7 +5400,7 @@ mod tests {
         )?;
 
-        assert_public_ty(&db, "src/a.py", "x", "Unbound | int");
+        assert_public_ty(&db, "src/a.py", "x", "int");
```
another case where it's important that we assert on may-be-unbound

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5466,7 +5506,7 @@ mod tests {
         )?;
 
-        assert_public_ty(&db, "src/a.py", "x", "Unbound | int");
+        assert_public_ty(&db, "src/a.py", "x", "int");
```
and another

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5544,7 +5592,7 @@ mod tests {
             "src/a.py",
             "x",
-            r#"Unbound | Literal[1] | Literal["a"] | Literal[b"foo"]"#,
+            r#"Literal[1] | Literal["a"] | Literal[b"foo"]"#,
```
we should assert on maybe-unbound here

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5573,7 +5621,7 @@ mod tests {
             &["Object of type 'NotIterable' is not iterable."],
         );
-        assert_public_ty(&db, "src/a.py", "x", "Unbound | Unknown");
+        assert_public_ty(&db, "src/a.py", "x", "Unknown");
```
and here

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5725,7 +5770,7 @@ mod tests {
         // and the inferred type will just be `BaseExceptionGroup` --Alex
         //
```
This comment (including the two previous lines GH won't let me comment on) can be removed.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5751,7 +5796,7 @@ mod tests {
         // and the inferred type will just be `BaseExceptionGroup` --Alex
         //
```
And this one can be removed too

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -93,6 +100,24 @@ pub(crate) fn global_symbol_ty<'db>(db: &'db dyn Db, file: File, name: &str) ->
     symbol_ty(db, global_scope(db, file), name)
 }
 
+/// Shorthand for `symbol_ty_by_id` that looks up a global symbol from the context of being in that
+/// module.
```
```suggestion
/// Shorthand for `symbol_ty_by_id` that looks up a global symbol from the context of being in that
/// module, including fallback to globals if (possibly) not defined.
```

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -93,6 +100,24 @@ pub(crate) fn global_symbol_ty<'db>(db: &'db dyn Db, file: File, name: &str) ->
+                db,
+                global_scope(db, file),
+                symbol,
+                SymbolTableLookupContext::GlobalWithinModule,
```
So I realized that a simpler approach that doesn't require threading this context through would be to just use `symbol_ty_by_id` (which can always use `Never` and never `Unbound` or `Unknown` as the unbound-ty), and then explicitly use the use-def map here to check directly whether the name may be undeclared and may be unbound, and then if so, look up the builtin type and union that with the symbol public type.

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -93,6 +100,24 @@ pub(crate) fn global_symbol_ty<'db>(db: &'db dyn Db, file: File, name: &str) ->
 
+/// Shorthand for `symbol_ty_by_id` that looks up a global symbol from the context of being in that
+/// module.
+pub(crate) fn global_symbol_lookup<'db>(db: &'db dyn Db, file: File, name: &str) -> Type<'db> {
```
I still think it might be fine to inline this into `TypeInferenceBuilder::lookup_name`, since I don't think it will have any other call sites? If we do keep it as a separate function, then I think it should include the builtins-fallback part too; that's part of a "global symbol lookup".

Also I don't think it needs to be `pub(crate)`.

carljm reviewed (changes requested):
Nice work! You did what I asked for yesterday, but I now realize that in some cases I asked for the wrong thing 😆 Sorry about that! I've talked this over with the team (by which I mean @AlexWaygood in this case) and I think we've got a clearer plan in mind now; let me know if any of the comments are not clear.

Thanks for all your work on this! You definitely didn't pick an easy first task :)

carljm commented:
Hi @pilleye! I'd love to get this landed soon since it affects other work we want to do. Really appreciate your work on it so far, and if you don't have time to wrap it up that's no problem, we can pick it up. Let me know if we should go ahead and do that!

pilleye commented:
Ah sorry, I've just been a bit busy the past week with other work, I'll try to get it in a ready-to-review state tonight.

pilleye commented:
I'm stuck on a couple tests.

I also have another question, why don't we use `symbol_ty(...)` in `infer.rs` for this line:

```
let member_ty = module_ty.member(...)
```

carljm commented:
> I'm stuck on a couple tests.

Taking a look! Thanks for all your work on this. If the changes to get everything passing look reasonably small, I might just go ahead and push them and land this, in the interest of efficiency, hope that's OK with you.

> I also have another question, why don't we use `symbol_ty(...)` in `infer.rs` for this line:
> 
> ```
> let member_ty = module_ty.member(...)
> ```

We kinda do use `global_symbol_ty`, because that's the entire implementation of `Type::member` for `Type::Module` types. It's just a layer of abstraction, since semantically what is happening here is an attribute access on a module object, and we model attribute accesses via `Type::member`.

pilleye commented:
I think I have it mostly working, there are just a few weird issues that feel orthogonal like:
```
---- types::infer::tests::exception_handler_with_invalid_syntax stdout ----
thread 'types::infer::tests::exception_handler_with_invalid_syntax' panicked at crates/red_knot_python_semantic/src/types/infer.rs:3519:9:
assertion `left == right` failed
  left: ["Object of type `Literal[reveal_type, reveal_type]` is not assignable to `Literal[reveal_type, reveal_type]`", "Revealed type is `Unknown`"]
 right: ["Revealed type is `Unknown`"]
```

I'm guessing this is because when we're looking at typeshed, we're replacing a union containing `Unknown` with a union containing `Unbound` and that is getting replaced with `Never` and removed from the union. So we were originally assigning something to `Any`/`Unknown`, which was fine, but now we're assigning literals to literals.

carljm commented:
@pilleye Ok, made some updates here and I think I got everything working -- didn't really change that much, but I split out the commits to help clarify what I changed and why, feel free to look at the commits if you're curious, lmk if anything doesn't make sense.

Now getting these weird errors about the Salsa crate from CI, need to figure out what's up with that but it doesn't seem related to this PR...

carljm commented:
Yeah ruff CI is currently broken for all PRs: https://github.com/astral-sh/ruff/pull/13780

pilleye commented:
> @pilleye Ok, made some updates here and I think I got everything working -- didn't really change that much, but I split out the commits to help clarify what I changed and why, feel free to look at the commits if you're curious, lmk if anything doesn't make sense.

That makes so much more sense! Appreciate the breakdown of commits :)

carljm commented:
This ends up being kind of a hacky approach to the original problem, in that we still keep Unbound in all the base type queries (`global_symbol_ty`, etc) but replace it with Never in unions wherever a type "escapes" as a public type. But that's fine; it gets the tests reflecting the types we actually should be seeing, which is progress. The better implementation will ultimately involve totally removing `Type::Unbound` anyway, which is a bigger change (https://github.com/astral-sh/ruff/issues/13671).

Thanks again @pilleye for all your work on this (very confusing, as it turned out) issue!
